exasol_advanced_analytics_framework/udf_communication/peer_communicator/background_thread/connection_closer/connection_is_closed_sender.py

```diff
@@ -0,0 +1,97 @@
+from enum import IntFlag, auto
+
+import structlog
+from structlog.typing import FilteringBoundLogger
+
+from exasol_advanced_analytics_framework.udf_communication import messages
+from exasol_advanced_analytics_framework.udf_communication.connection_info import ConnectionInfo
+from exasol_advanced_analytics_framework.udf_communication.peer import Peer
+from exasol_advanced_analytics_framework.udf_communication.peer_communicator.timer import Timer
+from exasol_advanced_analytics_framework.udf_communication.serialization import serialize_message
+from exasol_advanced_analytics_framework.udf_communication.socket_factory.abstract import Socket
+
+LOGGER: FilteringBoundLogger = structlog.get_logger()
+
+
+class _States(IntFlag):
+    INIT = auto()
+    RECEIVED_CLOSE_CONNECTION = auto()
+    RECEIVED_ACKKNOWLEDGE_CLOSE_CONNECTION = auto()
+    FINISHED = auto()
+
+
+class ConnectionIsClosedSender:
+    def __init__(self,
+                 out_control_socket: Socket,
+                 peer: Peer,
+                 my_connection_info: ConnectionInfo,
+                 timer: Timer):
+        self._timer = timer
+        self._peer = peer
+        self._out_control_socket = out_control_socket
+        self._states = _States.INIT
+        self._logger = LOGGER.bind(
+            peer=self._peer.dict(),
+            my_connection_info=my_connection_info.dict(),
+        )
+        self._logger.debug("init")
+
+    def received_close_connection(self):
+        self._logger.debug("received_close_connection", states=self._states)
+        self._states |= _States.RECEIVED_CLOSE_CONNECTION
+        self._timer.reset_timer()
+
+    def received_acknowledge_close_connection(self):
+        self._logger.debug("received_acknowledge_close", states=self._states)
+        self._states |= _States.RECEIVED_ACKKNOWLEDGE_CLOSE_CONNECTION
+
+    def try_send(self):
+        self._logger.debug("try_send", states=self._states)
+        should_we_send = self._should_we_send()
+        if should_we_send:
+            self._states |= _States.FINISHED
+            self._send_connection_is_closed_to_frontend()
+
+    def _should_we_send(self):
+        is_time = self._timer.is_time()
+        send_time_dependent = _States.RECEIVED_CLOSE_CONNECTION in self._states
+        send_time_independent = (
+                _States.RECEIVED_CLOSE_CONNECTION in self._states
```

Comment:
This is the only line that really changed compared to the ConnectionEstablisher, because for the ConnectionEstablisher it only counts if the other Peers does know us. If the other peer receives a Synchronize message, it answers with a Synchronize and a Ack message, so we only need on of it. However, for the ConnectionCloser, we need that both sides send CloseConnection, because otherwise we don't know it the other one wants to still send something to us.